File: cpp/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode *buildTree(vector<int> &inorder, vector<int> &postorder) {
        // Note: The Solution object is instantiated only once and is reused by each test case.
        return solve(inorder, 0, postorder, 0, inorder.size());
    }
    TreeNode *solve(vector<int> &inorder, int startI, vector<int> &postorder, int startP, int len){
        if (len == 0) {
            return NULL;
        }
        int rootVal = postorder[startP + len - 1];
        TreeNode *root = new TreeNode(rootVal);
        int rootOffset = 0;
        for (int n = 0; n < len; n ++) {
            if (inorder[startI + n] == rootVal){
                rootOffset = n;
                break;
            }
        }
        root->left = solve(inorder, startI, postorder, startP, rootOffset);
        root->right = solve(inorder, startI+rootOffset+1, postorder, startP+rootOffset, len-rootOffset-1);
        return root;
    }
};

File: Java/Construct_Binary_Tree_from_Inorder_and_Postorder_Traversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return buildTreeRecursively(inorder, 0, postorder, 0, inorder.length);
    }
    private TreeNode buildTreeRecursively(int[] inorder, int startI, int[] postorder, int startP, int n){
        if (n == 0) return null;
        TreeNode ret = new TreeNode(postorder[startP+n-1]);
        int offset = 0;
        while (offset < n) {
            if (inorder[startI+offset] == postorder[startP+n-1]) break;
            offset ++;
        }
        ret.left = buildTreeRecursively(inorder, startI, postorder, startP, offset);
        ret.right = buildTreeRecursively(inorder, startI+offset+1, postorder, startP+offset, n-offset-1);
        return ret;
    }
}
